N°5775 - Allow configuration of OAuth client on MS Azure with single tenant

delassiter opened the pull request:
When using OAuth2 to authenticate to Microsoft Azure in order to access a mail account, currently only multi-tenant configurations are possible. This is often forbidden by company policy in large corporations.

I added a "tenant" field to the OAuth-Client, which defaults to "common", and pass it to the `OAuthClientProviderAzure`.

delassiter committed:
Add Tenant-Support for Azure OAuthClient

Hipska commented on datamodels/2.x/itop-oauth-client/datamodel.itop-oauth-client.xml:
```diff
@@ -380,6 +380,11 @@ HTML
+        <field id="tenant" xsi:type="AttributeString">
+          <sql>tenant</sql>
+          <default_value>common</default_value>
+          <is_null_allowed>true</is_null_allowed>
```
Maybe make this a required field, as the default is already `common`?

delassiter replied:
I've changed the "is_null_allowed" to false.

piRGoif commented:
Thanks Denis ! 
Warning, the N°XXXX is a reference to a Combodo ticket
The multi tenant handling was asked in SF ticket [#2107](https://sourceforge.net/p/itop/tickets/2107/) indeed, but it was referenced as N°5775 in Combodo internal DB.

delassiter committed:
Improvement: Make tenant required

delassiter commented:
@piRGoif, Sorry about the mixup with the Ticket-ID

delassiter committed:
Improvment: Removed check for null-value

Since last commit, the "tenant"-field either set to a custom value or "common" by default. It is not allowed to be null

piRGoif commented:
> Sorry about the mixup with the Ticket-ID

Absolutly no problem, thanks for your contribution !

dutch2005 commented:
When will this proposed edit be made available in production version of this plugin?

piRGoif commented:
> When will this proposed edit be made available in production version of this plugin?

Hello,
The PR hasn't been reviewed yet, so we cannot answer for now.

Molkobain commented on datamodels/2.x/itop-oauth-client/en.dict.itop-oauth-client.php:
```diff
@@ -93,6 +93,7 @@
 	'Class:OAuthClientAzure/Attribute:used_for_smtp+' => 'At least one OAuth client must have this flag to “Yes”, if you want iTop to use it for sending mails',
 	'Class:OAuthClientAzure/Attribute:used_for_smtp/Value:yes' => 'Yes',
 	'Class:OAuthClientAzure/Attribute:used_for_smtp/Value:no' => 'No',
+    'Class:OAuthClientAzure/Attribute:tenant' => 'Tenant',
```
```suggestion
    'Class:OAuthClientAzure/Attribute:tenant' => 'Tenant',
    'Class:OAuthClientAzure/Attribute:tenant+' => '~~',
```

Add a description entry that will be used as a tooltip to help understand where to get it and what kind of value to set.

Molkobain reviewed (commented):
_PR will be tested next monday to confirm that this way of handling single-tenant is ok._

Molkobain commented:
Tested with Stephane and Stephen today, works great! 🙌

Things left to do:
  - Add a tooltip on the field to explain where to retrieve the tenant ID / name (works with both)
  - Improve error catching to display a message when a wrong tenant is provided, otherwise it fails silently

_Example of log in case of wrong tenant_
```
2023-12-11 15:27:12 | Debug   | 1     | GetAuthorizationUrl for OAuthClientAzure::3 | OAuth |||
2023-12-11 15:27:13 | Error   | 1     | invalid_tenant
AADSTS90002: Tenant 'abcdefg' not found. Check to make sure you have the correct tenant ID and are signing into the correct cloud. Check with your subscription administrator, this may happen if there are no active subscriptions for the tenant. Trace ID: 123456-1234-1234-1234-123456789 Correlation ID: 1234567789-1234-1234-1234-123456789 Timestamp: 2023-12-11 14:27:12Z | IssueLog |||
```

Molkobain commented:
Accepted during functional review:
  - In case of incorrect tenant ID, no error is displayed at the moment, Combodo will handle that part
  - Will be fixed in iTop 2.7.11, 3.1.2, 3.2.0

Molkobain commented:
@delassiter can you send me your fullname and postal address to guillaume.lajarige at combodo.com ? We would like to send you thankful stickers. :)

piRGoif commented:
2.7.10 and 3.1.1 versions were published, this PR can be merged !

Molkobain committed:
Add field description
